Set default value for fields: add defaulter processor

jregistr opened the pull request:
Hellos! 👋 

So I've been experimenting w/ this amazing application y'all have here. 
Now, for certain fields coming out of the grok parser Input, I wanted a processor to set a default value when they're either nil, empty or zero. 

I didn't see a way to do this with the current set of processors so I got to some experimenting and made this very rough implementation.

I made this draft PR to get some thoughts on if a processor is the way to approach this, as well as thoughts on the config and implementation.

All comments are welcomed. 

Thank you all very much for taking a look.

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

danielnelson commented:
Thanks for the pull request, I'm sure this will come in handy.

How about this as an alternate configuration style, I think it would fit more closely with existing plugins:
```toml
[[processors.default]]
  [processors.default.fields]
    foo = "bar"
    time_idle = 0
```

Operating on metric's with specific names and tags can be done with the [metric filtering](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#metric-filtering) options that are available automatically for all plugins:

```toml
[[processors.default]]
  namepass = ["cpu"]
  [processors.default.fields]
    time_idle = 0
```

We should make sure to support floats/ints/bool/strings for the field type.

related #2564

jregistr commented:
I'm still pretty new to golang... I'm not sure why the tidy step is failing. I ran it locally and checked the changes.

jregistr commented:
But besides that, this is ready for a look. I updated the config according to your suggestion @danielnelson . I went ahead and updated the impl and tests to match.

danielnelson commented on plugins/processors/defaulter/README.md:
```diff
@@ -0,0 +1,43 @@
+# Defaulter Processor
```
Let's call the plugin either `defaults` or `default`, this fits the noun style of most of our processors.

jregistr replied:
Gotcha, `defaults` is a great replacement!

danielnelson commented on plugins/processors/defaulter/README.md:
````diff
@@ -0,0 +1,43 @@
+
+```diff
+- lb,http_method=GET cache_status=HIT latency=230
++ lb,http_method=GET cache_status latency=230 status_code="N/A"
````
In order for these examples to be valid line protocol we must separate fields with a comma:
```
lb,http_method=GET cache_status,latency=230,status_code="N/A"
```

danielnelson commented on plugins/processors/defaulter/defaulter.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+const sampleConfig = `
+	## Ensure's a set of fields always exist on your metric(s) with at the specified default value.
+	## For any given field pair (key = default) under *[processors.defaulter.fields]*, if it's not set, or its value
+	## is 'empty' on the incoming metric, a new or updated metric is set on the metric with the specified default.
+	## 
+	#	[processors.defaulter.fields]
+	#	 field_1 = "bar"
+	#    time_idle = 0
+	#    is_error = true
```
Follow this guide to match style with other plugins: https://github.com/influxdata/telegraf/wiki/SampleConfig

2 space indent, wrap line length, etc.

danielnelson commented on plugins/processors/defaulter/defaulter.go:
```diff
@@ -0,0 +1,62 @@
+	for _, metric := range inputMetrics {
+		for defField, defValue := range def.DefaultFieldsSets {
+			if maybeCurrent, isSet := metric.GetField(defField); !isSet {
+				def.Log.Debugf("Field with name: %v, was not set on metric: %v.", defField, metric.Name())
```
Remove these debug log messages, they are a bit too verbose for the final code, also on L49

danielnelson commented on plugins/processors/defaulter/defaulter.go:
```diff
@@ -0,0 +1,62 @@
+			if maybeCurrent, isSet := metric.GetField(defField); !isSet {
+				def.Log.Debugf("Field with name: %v, was not set on metric: %v.", defField, metric.Name())
+				metric.AddField(defField, defValue)
+			} else if maybeCurrent == "" || maybeCurrent == " " {
```
Call `strings.TrimSpace` and compare it against "" when deciding if the value needs replaced.  Undate the note in the README about what is considered unset.

danielnelson commented on plugins/processors/defaulter/defaulter.go:
```diff
@@ -0,0 +1,62 @@
+				metric.AddField(defField, defValue)
+			} else if maybeCurrent == "" || maybeCurrent == " " {
+				def.Log.Debugf("Field with name: %v was set, but the value (%v) is considered empty. Setting new value to %v", defField, maybeCurrent, defValue)
+				metric.RemoveField(defField)
+				metric.AddField(defField, defValue)
```
You can just call AddField, it will replace the existing value.

jregistr replied:
aaah, interesting! Will update!

jregistr commented:
@danielnelson  Thank you for the feedback, I'll update and push the changes.

jregistr commented:
Hello again @danielnelson, I believe this is ready for another look.

jregistr commented:
>Error: No such image: quay.io/influxdb/telegraf-ci:1.13.8

Looks like ci failed to download an image?

reimda commented on plugins/processors/defaults/README.md:
```diff
@@ -0,0 +1,42 @@
+
+1. The field is nil on the incoming metric
+1. The field is not nil, but its value is an empty string.
+1. The field is not nil, but it's value is a string of one or more empty spaces.
```
typo: it's should be its

jregistr replied:
Good catch! Fixed!

jregistr commented:
Amazing!
